library/core/src/num/dec2flt/common.rs

```diff
@@ -0,0 +1,220 @@
+
+    /// Get the slice after skipping all leading characters equal c1 or c2.
+    fn skip_chars2(&self, c1: u8, c2: u8) -> &[u8] {
+        let mut s = self.as_ref();
+        while s.first_is2(c1, c2) {
+            s = s.advance(1);
+        }
+        s
+    }
+
+    /// Read 8 bytes as a 64-bit integer in little-endian order.
+    unsafe fn read_u64_unchecked(&self) -> u64 {
+        debug_assert!(self.check_len(8));
+        let src = self.as_ref().as_ptr() as *const u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        u64::from_le(unsafe { ptr::read_unaligned(src) })
+    }
+
+    /// Try to read the next 8 bytes from the slice.
+    fn read_u64(&self) -> Option<u64> {
+        if self.check_len(8) {
+            // SAFETY: self must be at least 8 bytes.
+            Some(unsafe { self.read_u64_unchecked() })
+        } else {
+            None
+        }
+    }
+
+    /// Calculate the offset of slice from another.
+    unsafe fn offset_from(&self, other: &Self) -> isize {
+        // SAFETY: safe as long as self and other are of the same array.
+        unsafe { self.as_ref().as_ptr().offset_from(other.as_ref().as_ptr()) }
+    }
+}
+
+impl ByteSlice for [u8] {}
+
+/// Helper methods to process mutable bytes.
+pub trait ByteSliceMut: ByteSlice + AsMut<[u8]> {
+    /// Write a 64-bit integer as 8 bytes in little-endian order.
+    unsafe fn write_u64_unchecked(&mut self, value: u64) {
+        debug_assert!(self.as_ref().len() >= 8);
+        let dst = self.as_mut().as_mut_ptr() as *mut u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        unsafe {
+            ptr::write_unaligned(dst, u64::to_le(value));
+        }
+    }
+}
+
+impl ByteSliceMut for [u8] {}
+
+/// Bytes wrapper with specialized methods for ASCII characters.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct AsciiStr<'a> {
+    slc: &'a [u8],
+}
+
+impl<'a> AsciiStr<'a> {
+    pub fn new(slc: &'a [u8]) -> Self {
+        Self { slc }
+    }
+
+    /// Advance the view by n, advancing it in-place to (n..).
+    pub unsafe fn step_by(&mut self, n: usize) -> &mut Self {
+        // SAFETY: safe as long n is less than the buffer length
+        self.slc = unsafe { self.slc.get_unchecked(n..) };
+        self
+    }
+
+    /// Advance the view by n, advancing it in-place to (1..).
+    pub unsafe fn step(&mut self) -> &mut Self {
+        // SAFETY: safe as long as self is not empty
+        unsafe { self.step_by(1) }
+    }
+
+    /// Iteratively parse and consume digits from bytes.
+    pub fn parse_digits(&mut self, mut func: impl FnMut(u8)) {
+        // SAFETY: self cannot be empty
+        unsafe {
+            while self.first_isdigit() {
+                func(self.first_unchecked() - b'0');
+                self.step();
+            }
+        }
+    }
+}
+
+impl<'a> AsRef<[u8]> for AsciiStr<'a> {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.slc
+    }
+}
+
+impl<'a> ByteSlice for AsciiStr<'a> {}
```

Comment:
If you wanted, you could avoid this boilerplate on each type by adding a blanket `impl<T: AsRef<[u8]>> ByteSlice for T {}`. I have no preference on which is easier.

Reply:
This added a lot of errors due to lifetime and borrow issues, which could be fixable likely with some careful rework, but I think the existing implementation is easier and more readable.